File: agent-go/internal/headless/headless.go
// Package headless implements the PI_HEADLESS=1 test protocol.
//
// Input (stdin, JSON):
//
//	{"prompt": "...", "mock_turns": [...] | null, "cwd": "..."}
//
// Output (stdout, NDJSON):
//
//	{"type": "agent_start"}
//	{"type": "turn_start",  "turn": 1}
//	{"type": "tool_call",   "id": "t1", "name": "read",   "args": {...}}
//	{"type": "tool_result", "id": "t1", "name": "read",   "content": "...", "is_error": false}
//	{"type": "turn_end",    "turn": 1}
//	{"type": "agent_end",   "turns": 2}
//	{"type": "error",       "message": "..."}   (only on fatal errors)
package headless

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"os"

	"github.com/pi-agent/agent-go/internal/agent"
	"github.com/pi-agent/agent-go/internal/ai"
	"github.com/pi-agent/agent-go/internal/ai/mock"
	"github.com/pi-agent/agent-go/internal/tools"
)

// Input is the JSON payload read from stdin.
type Input struct {
	Prompt     string           `json:"prompt"`
	MockTurns  []mock.TurnSpec  `json:"mock_turns"`
	CWD        string           `json:"cwd"`
	HasMock    bool             // set internally: mock_turns key was present
}

// UnmarshalJSON handles the distinction between null and absent mock_turns.
func (inp *Input) UnmarshalJSON(data []byte) error {
	type Alias Input
	var raw struct {
		Alias
		MockTurns *[]mock.TurnSpec `json:"mock_turns"`
	}
	if err := json.Unmarshal(data, &raw); err != nil {
		return err
	}
	*inp = Input(raw.Alias)
	if raw.MockTurns != nil {
		inp.MockTurns = *raw.MockTurns
		inp.HasMock = true
	}
	return nil
}

const systemPrompt = `You are a coding assistant. You have access to tools to read, write, ` +
	`and edit files, run bash commands, search for files and content, and list directories.

Guidelines:
- Use read to examine files before editing them.
- Use edit (with the edits array) instead of write when making targeted changes.
- Use bash for running tests, builds, and shell commands.
- Work methodically. When you finish, summarise what you did.`

// Run reads stdin, executes the agent, writes NDJSON events to stdout.
func Run(apiKey, providerName, modelID string) error {
	// Parse input from stdin.
	rawBytes, err := io.ReadAll(os.Stdin)
	if err != nil {
		return fmt.Errorf("headless: read stdin: %w", err)
	}
	var inp Input
	if err := json.Unmarshal(rawBytes, &inp); err != nil {
		return fmt.Errorf("headless: parse stdin: %w", err)
	}

	// Resolve working directory.
	cwd := inp.CWD
	if cwd == "" {
		var err error
		cwd, err = os.Getwd()
		if err != nil {
			cwd = "."
		}
	}

	// Build provider.
	var prov ai.Provider
	if inp.HasMock {
		prov = mock.New(inp.MockTurns)
	} else {
		prov = buildRealProvider(providerName, apiKey)
	}

	// Build agent context.
	agentCtx := &agent.Context{
		SystemPrompt: systemPrompt,
		Messages:     nil,
		Tools:        buildTools(cwd),
		ModelID:      modelID,
		Provider:     providerName,
		APIKey:       apiKey,
	}

	// Override max turns from env.
	cfg := agent.DefaultConfig()
	if v := os.Getenv("PI_MAX_TURNS"); v != "" {
		var n int
		if _, err := fmt.Sscanf(v, "%d", &n); err == nil && n > 0 {
			cfg.MaxTurns = n
		}
	}

	// Run agent loop and emit NDJSON events.
	ctx := context.Background()
	eventCh := agent.Run(ctx, inp.Prompt, agentCtx, prov, cfg)

	emit := func(v any) {
		b, _ := json.Marshal(v)
		fmt.Println(string(b))
	}

	turn := 0
	for ev := range eventCh {
		switch ev.Type {
		case agent.EventAgentStart:
			emit(map[string]any{"type": "agent_start"})

		case agent.EventTurnStart:
			turn++
			emit(map[string]any{"type": "turn_start", "turn": turn})

		case agent.EventTurnEnd:
			emit(map[string]any{
				"type":          "turn_end",
				"turn":          turn,
				"input_tokens":  ev.InputTokens,
				"output_tokens": ev.OutputTokens,
			})

		case agent.EventStreamChunk:
			// Stream chunks are noisy; omit from headless output to keep it clean.
			// TUI uses them; tests don't need them.

		case agent.EventToolExecStart:
			emit(map[string]any{
				"type": "tool_call",
				"id":   ev.ToolCallID,
				"name": ev.ToolName,
				"args": ev.ToolArgs,
			})

		case agent.EventToolExecUpdate:
			// Partial updates — skip in headless mode.

		case agent.EventToolExecEnd:
			content := ""
			if ev.ToolResult != nil {
				content = ev.ToolResult.Content
			}
			emit(map[string]any{
				"type":     "tool_result",
				"id":       ev.ToolCallID,
				"name":     ev.ToolName,
				"content":  content,
				"is_error": ev.IsError,
			})

		case agent.EventAgentEnd:
			emit(map[string]any{
				"type":         "agent_end",
				"turns":        turn,
				"error_message": ev.ErrorMessage,
			})

		case agent.EventError:
			emit(map[string]any{
				"type":    "error",
				"message": ev.ErrorMessage,
			})
		}
	}

	return nil
}

func buildTools(cwd string) []tools.Tool {
	return []tools.Tool{
		tools.NewReadTool(cwd),
		tools.NewWriteTool(cwd),
		tools.NewEditTool(cwd),
		tools.NewBashTool(cwd),
		tools.NewFindTool(cwd),
		tools.NewGrepTool(cwd),
		tools.NewLsTool(cwd),
	}
}

File: agent-python/src/pi/headless.py
# src/pi/headless.py
"""
Headless mode: PI_HEADLESS=1

Reads JSON from stdin, runs the agent, emits NDJSON events to stdout.

Input (stdin):
    {"prompt": "...", "mock_turns": [...] | null, "cwd": "..."}

Output (stdout NDJSON):
    {"type": "agent_start"}
    {"type": "turn_start",  "turn": 1}
    {"type": "tool_call",   "id": "t1", "name": "read",   "args": {...}}
    {"type": "tool_result", "id": "t1", "name": "read",   "content": "...", "is_error": false}
    {"type": "turn_end",    "turn": 1}
    {"type": "agent_end",   "turns": 2}
    {"type": "error",       "message": "..."}
"""
from __future__ import annotations

import asyncio
import json
import os
import sys
from typing import Any

from .agent.context import AgentContext, AgentConfig
from .agent.events import (
    AgentEnd, AgentError, AgentStart,
    ToolExecEnd, ToolExecStart, ToolExecUpdate,
    TurnEnd, TurnStart,
)
from .agent.loop import agent_loop
from .tools.base import Tool
from .tools.read import ReadTool
from .tools.write import WriteTool
from .tools.edit import EditTool
from .tools.bash import BashTool
from .tools.find import FindTool
from .tools.grep import GrepTool
from .tools.ls import LsTool

SYSTEM_PROMPT = (
    "You are a coding assistant. You have access to tools to read, write, and edit files, "
    "run bash commands, search for files and content, and list directories.\n\n"
    "Guidelines:\n"
    "- Use read to examine files before editing them.\n"
    "- Use edit (with the edits array) instead of write when making targeted changes.\n"
    "- Use bash for running tests, builds, and shell commands.\n"
    "- Work methodically. When you finish, summarise what you did."
)


def _build_tools(cwd: str) -> list[Tool]:
    # Python tools resolve paths relative to os.getcwd(),
    # so we chdir to the workspace before running.
    # The cwd is applied in _run() via os.chdir().
    return [
        ReadTool(),
        WriteTool(),
        EditTool(),
        BashTool(),
        FindTool(),
        GrepTool(),
        LsTool(),
    ]


def _emit(obj: dict) -> None:
    print(json.dumps(obj, ensure_ascii=False), flush=True)


def _make_mock_provider(mock_turns: list[dict]):
    """Return a mock LLMProvider that replays the given turns."""
    from .ai.base import LLMProvider, StreamOptions
    from .ai.types import (
        AssistantMessage, StreamDone, TextDelta,
        ToolCall, ToolCallEnd, ToolCallStart,
    )

    _EMPTY_ASSISTANT = AssistantMessage(content=[])

    class MockProvider(LLMProvider):
        def __init__(self, turns: list[dict]) -> None:
            self._turns = turns
            self._idx = 0

        async def stream(self, messages, system_prompt, tools, opts):
            if self._idx >= len(self._turns):
                yield StreamDone(stop_reason="stop", message=_EMPTY_ASSISTANT)
                return

            turn = self._turns[self._idx]
            self._idx += 1

            if text := turn.get("text"):
                yield TextDelta(index=0, delta=text, partial=_EMPTY_ASSISTANT)

            for i, tc in enumerate(turn.get("tool_calls", [])):
                tc_id = tc.get("id") or f"mock-{self._idx}-{i}"
                yield ToolCallStart(index=i, partial=_EMPTY_ASSISTANT)
                yield ToolCallEnd(
                    index=i,
                    partial=_EMPTY_ASSISTANT,
                    tool_call=ToolCall(
                        id=tc_id,
                        name=tc["name"],
                        arguments=tc.get("args", {}),
                    ),
                )

            stop = turn.get("stop_reason", "end_turn")
            if stop in ("tool_use", "tool_calls"):
                yield StreamDone(stop_reason="tool_use", message=_EMPTY_ASSISTANT)
            else:
                yield StreamDone(stop_reason="stop", message=_EMPTY_ASSISTANT)

    return MockProvider(mock_turns)


async def _run(
    prompt: str,
    provider,
    cwd: str,
    api_key: str,
    model_id: str,
    max_turns: int,
) -> None:
    # Change to workspace directory so tools resolve relative paths correctly.
    os.chdir(cwd)
    tools = _build_tools(cwd)
    config = AgentConfig(
        api_key=api_key,
        model_id=model_id,
        max_turns=max_turns,
    )
    ctx = AgentContext(
        system_prompt=SYSTEM_PROMPT,
        messages=[],
        tools=tools,
        config=config,
    )

    turn = 0

    def emit(event: Any) -> None:
        nonlocal turn
        if isinstance(event, AgentStart):
            _emit({"type": "agent_start"})
        elif isinstance(event, TurnStart):
            turn += 1
            _emit({"type": "turn_start", "turn": turn})
        elif isinstance(event, TurnEnd):
            _emit({"type": "turn_end", "turn": turn})
        elif isinstance(event, ToolExecStart):
            _emit({"type": "tool_call", "id": event.id, "name": event.name, "args": event.args})
        elif isinstance(event, ToolExecUpdate):
            pass  # skip partial updates in headless mode
        elif isinstance(event, ToolExecEnd):
            _emit({
                "type": "tool_result",
                "id": event.id,
                "name": event.name,
                "content": event.result.content if event.result else "",
                "is_error": event.is_error,
            })
        elif isinstance(event, AgentEnd):
            _emit({"type": "agent_end", "turns": turn})
        elif isinstance(event, AgentError):
            _emit({"type": "error", "message": str(event.message)})

    await agent_loop(prompt, ctx, provider, emit)


def run() -> None:
    """Entry point for headless mode. Reads stdin, writes stdout."""
    raw = sys.stdin.read()
    try:
        inp = json.loads(raw)
    except json.JSONDecodeError as exc:
        _emit({"type": "error", "message": f"invalid JSON input: {exc}"})
        sys.exit(1)

    prompt: str = inp.get("prompt", "")
    mock_turns = inp.get("mock_turns")   # None means real LLM
    cwd: str = inp.get("cwd") or os.getcwd()

    # API key — allow dummy when using mock provider.
    api_key = (
        os.environ.get("KIMI_API_KEY")
        or os.environ.get("MOONSHOT_API_KEY")
        or os.environ.get("ANTHROPIC_API_KEY")
        or os.environ.get("OPENROUTER_API_KEY")
        or os.environ.get("OPENAI_API_KEY")
        or "sk-test-dummy"
    )
    model_id = os.environ.get("PI_MODEL", "kimi-for-coding")
    max_turns = int(os.environ.get("PI_MAX_TURNS", "50"))

    if mock_turns is not None:
        provider = _make_mock_provider(mock_turns)
    else:
        provider_name = os.environ.get("PI_PROVIDER", "kimi")
        if provider_name == "openai":
            from .ai.openai import OpenAIProvider
            provider = OpenAIProvider()
        elif provider_name == "kimi":
            from .ai.openai import KimiProvider
            provider = KimiProvider()
        elif provider_name == "openrouter":
            from .ai.openai import OpenRouterProvider
            provider = OpenRouterProvider()
        else:
            from .ai.anthropic import AnthropicProvider
            provider = AnthropicProvider()

    asyncio.run(_run(prompt, provider, cwd, api_key, model_id, max_turns))
